### PR TITLE
Add support for I2C master driver and fixed a spelling error

### PR DESCRIPTION
**Pull request by PurpleCloudX**
* Added support for the new I2C Master driver.
* Updated the example in the file.
* Fixed a spelling error: BMX280_ADDERSS_HI -> BMX280_ADDRESS_HI

### CODE REVIEW

File: include/bmx280.h [hunk 1]
```diff
@@ -16,24 +16,42 @@ extern "C" {
...
  * @return A non-null pointer to the driver structure on success.
  */
 BMXAPI bmx280_t* bmx280_create(i2c_port_t port);
+#else
+/**
+ * Create an instance of the BMX280 driver.
+ * @param bus_handle The I2C master handle via port.
+ * @return A non-null pointer to the driver structure on success.
+ */
+BMXAPI bmx280_t* bmx280_create(i2c_master_bus_handle_t bus_handle);
```
Comment (utkumaden) [inline include/bmx280.h]: I think the review I made in PR #4 applies here as well. I would rather not create a situation where people can accidentally compile the BMX280 driver with the incorrect I2C driver mode. I suggest you refactor this like so:

```c
#include <assert.h>

#if CONFIG_USE_I2C_NEW_DEVICE
// Your new entry point.
BMXAPI bmx280_t bmx280_create_master(i2c_master_dev_handle_t dev_handle);
#define bmx280_create_legacy(port) static_assert(0, "You have the wrong driver configuration for using the legacy I2C driver.")
#define bmx280_create(port) bmx280_create_legacy(port)
#else
// The original entry point but renamed.
BMXAPI bmx280_t bmx280_create_legacy(i2c_port_t port);
// legacy define for existing code bases
#define bmx280_create(port) (bmx280_create_legacy((port)))
#define bmx280_create_master(port) static_assert(0, "You have the wrong driver configuration for using the new I2C master driver.")
#endif
```

It brings up the C language version from C99 to C11 but I think this is important for legacy users.

File: Kconfig [hunk 2]
```diff
@@ -1,4 +1,28 @@
 menu "BMX280 Options"
+    choice I2C_DRIVER_SETTING
+        prompt "I2C driver setting"
+        help
+            Select I2C driver: I2C_OLD: Use old driver I2C.h, I2C_NEW: Use new i2c_master.h(ESP-IDF >=5.3)
+        default USE_I2C_OLD_DEVICE
```
Comment (utkumaden) [inline Kconfig]: I don't like the new terminology you have introduced here. New/Old might imply that the BMP280/BME280 devices being used are new and old.

I suggest you rename these config keys and their help strings as I2C Master and I2C Legacy.

File: bmx280.c [hunk 3]
```diff
@@ -69,11 +69,28 @@
 // Value of REG_CHPID for BMP280 (Production)
 #define BMP280_ID2 0x58
 
+#if CONFIG_USE_I2C_NEW_DEVICE
+// I2C Master configuration structure (temporary)
+typedef struct i2c_config_t{
```
Comment (utkumaden) [inline bmx280.c]: I suggest you merge this into the main struct. I don't see a reason to have this be its own type.

File: bmx280.c [hunk 4]
```diff
@@ -309,6 +397,7 @@ static esp_err_t bmx280_calibrate(bmx280_t *bmx280)
     return ESP_OK;
 }
 
+#if !(CONFIG_USE_I2C_NEW_DEVICE)
```
Comment (utkumaden) [inline bmx280.c]: This macro sequence is a bit wack, I think you should probably split this into two like I suggest in the following review comments.

File: bmx280.c [hunk 5]
```diff
@@ -318,13 +407,39 @@ bmx280_t* bmx280_create(i2c_port_t port)
...
+        bmx280_close(bmx280);
+        return NULL;
+    }
     return bmx280;
 }
 
 void bmx280_close(bmx280_t *bmx280)
 {
+    #if CONFIG_USE_I2C_NEW_DEVICE
+    free(bmx280->i2c_cfg);
```
Comment (utkumaden) [inline bmx280.c]: I suggest you use an if NULL check around the deconstructor for the properties. It's legal to free(NULL) but not legal to dereference a field.

**Review by utkumaden — CHANGES_REQUESTED** [review]
Hello, thank you for taking your time to contribute. I have a few comments/changes I would like you to address before I can merge this.

Do note that a similar pull request is pending (#4).

File: bmx280.c [hunk 6]
```diff
@@ -122,7 +139,27 @@ struct bmx280_t{
...
+ * Read from sensor.
+ * @param bmx280 Driver Sturcture.
+ * @param dev_addr Chip addresses.
+ */
+static void bmx280_device_create(bmx280_t *bmx280, const uint16_t dev_addr)
+{
+    ESP_LOGI("bmx280", "device_create for BMP280/BME280 sensors on ADDR %X", dev_addr);
+    bmx280->i2c_cfg->dev_cfg.device_address = dev_addr;
+    // Add device to the I2C bus
+    ESP_ERROR_CHECK(i2c_master_bus_add_device(bmx280->i2c_cfg->bus_handle, &bmx280->i2c_cfg->dev_cfg, &bmx280->i2c_dev));
```
Comment (utkumaden) [inline bmx280.c]: Don't assert for the user at runtime, I suggest you bubble this error up by returning the error value.

**Comment by PurpleCloudX** [comment]
Thank you very much for your feedback and suggestions on my code. This is my first time submitting a pull request, and your guidance is incredibly helpful. I will make the suggested changes to the code later today or tomorrow.

File: bmx280.c (hunk 3 above)
Comment (PurpleCloudX) [reply bmx280.c]: Thank you for your suggestion! I understand your perspective, I think the contents of this struct are not used again after the dev_handler is successfully created. By keeping it as a separate type and freeing it after use, we can save some memory.

Comment (utkumaden) [reply bmx280.c]: Not that a few bytes of memory really matters here, but then maybe stack memory is better for this? Does the information in this struct need to be used in bmx280_init?

Comment (PurpleCloudX) [reply bmx280.c]: No, it is not used in the bmx280_init function, it is only used in the bmx280_create,bmx280_probe,bmx280_device_create function to set chip speed and address

**Comment by PurpleCloudX** [comment]
Hello,Thank you for your valuable feedback and suggestions. I have made the requested changes to the code. If you have some time, could you please review it again? 
Best regards

File: bmx280.c [hunk 7]
```diff
@@ -309,22 +405,55 @@ static esp_err_t bmx280_calibrate(bmx280_t *bmx280)
...
+        return NULL;
+    }
+    return bmx280;
+}
+#endif
 
 void bmx280_close(bmx280_t *bmx280)
 {
+    #if CONFIG_USE_I2C_MASTER_DRIVER
+    if(bmx280->i2c_dev!=NULL)
```
Comment (utkumaden) [inline bmx280.c]: This still dereferences a pointer that may not be initialized (The `bmx280` driver structure as well).

```c
void bmx280_close(bmx280_t *bmx280)
{
    #if CONFIG_USE_I2C_MASTER_DRIVER
    if (bmx280 != NULL && bmx280->i2c_dev != NULL)
    {
        i2c_master_bus_rm_device(bmx280->i2c_dev);
    }
    #endif
    free(bmx280);
}
```

Notice I also check for the pointer bmx280 itself.

**Review by utkumaden — APPROVED** [review]
There is only one minor issue left, it would be nicer if you fixed it yourself. Otherwise I can fix it at merge time.

I initially thought #5 was a wontfix but you are correct, my macro definition is not correct. Whoops.

**Comment by utkumaden** [comment]
Thank you for taking your time. I've verified the new driver works on my hardware. Merged! :confetti_ball:

**Pull request merged**